railties/lib/minitest/rails_plugin.rb

```diff
@@ -39,8 +42,12 @@ def self.plugin_rails_options(opts, options)
   # minitest-reporters, maxitest and others.
   def self.plugin_rails_init(options)
     unless options[:full_backtrace] || ENV["BACKTRACE"]
-      # Plugin can run without Rails loaded, check before filtering.
-      Minitest.backtrace_filter = ::Rails.backtrace_cleaner if ::Rails.respond_to?(:backtrace_cleaner)
+      # Can use in a Rails plugin test suite or without Rails loaded.
+      if Pathname.pwd == ::Rails.try(:root)
```

Comment:
It really seems like this line should be:

```suggestion
      if ::Rails.respond_to?(:backtrace_cleaner) && !defined?(ENGINE_ROOT)
```

But as you [pointed out](https://github.com/rails/rails/pull/38443#issuecomment-603489997), `ENGINE_ROOT` is lost during `rake test` because Rake's test loader [spawns a new process](https://github.com/ruby/rake/blob/v13.0.1/lib/rake/testtask.rb#L117).

I'm tempted to suggest that we accept this limitation, especially since after https://github.com/rails/rails/pull/39304 the failure mode is an unfiltered backtrace, rather than no backtrace. It's possible to run a Rails application's tests from outside its root directory, which will confuse this check; messing with behaviour for applications to improve support for `rake test` from plugins seems like the wrong tradeoff to me.

Reply:
> But as you [pointed out](https://github.com/rails/rails/pull/38443#issuecomment-603489997), `ENGINE_ROOT` is lost during `rake test` because Rake's test loader [spawns a new process](https://github.com/ruby/rake/blob/v13.0.1/lib/rake/testtask.rb#L117).

Unfortunately, it's not only when running `rake test`.  `ENGINE_ROOT` is not set for simple railtie plugins.  It's [set by `bin/rails`](https://github.com/rails/rails/blob/04dcf56cec38fc91f7ac6694d8a4fad66dcdf4dd/railties/lib/rails/generators/rails/plugin/templates/bin/rails.tt#L4), which is [only generated for engine plugins](https://github.com/rails/rails/blob/04dcf56cec38fc91f7ac6694d8a4fad66dcdf4dd/railties/lib/rails/generators/rails/plugin/plugin_generator.rb#L161).  We could possibly set `ENGINE_ROOT` in `bin/test` or `test_helper.rb`, but I'm not sure that would be appropriate, and it wouldn't fix the issue for existing plugins.

> I'm tempted to suggest that we accept this limitation, especially since after #39304 the failure mode is an unfiltered backtrace

True, though the backtrace will be less pretty than expected, since it will include many lines that would normally be filtered by `Minitest::BacktraceFilter`.

> It's possible to run a Rails application's tests from outside its root directory, which will confuse this check

My thinking was that `ActiveSupport::BacktraceCleaner` *might* be more desirable in this case, since it would keep backtrace paths absolute, rather than make them relative to some other directory (`Rails.root`).

With all that said, I just pushed another new approach based on 0c8df42b50764833d5f84c1cb806aa6a84c70199 that uses `Rails.backtrace_cleaner` when applicable, but otherwise uses the default `Minitest.backtrace_filter`.

Reply:
Yes! That's even better ❤️